### ISSUE TITLE
Follow-up to Pocket background fetch

### ISSUE THREAD

**Issue by dnarcese** [mozilla-mobile/firefox-tv#2325]
Follow-up tasks to #2223 
### Vision statement / What / Requirements
The tasks that were not finished in the original PR:

- Remove remaining old Pocket foreground fetch code
  - [ ] Rename PocketEndpoint?
  - [ ] Remove PocketRepoCache and comment why it's not necessary (b/c we only content in onStart)
  - [ ]  Remove unused code
-  [ ] Clean up comments in scheduler

- Automated tests
  -  [ ] No API key
  -  [ ]  Non en-US locales
  - [ ]  Failure to hit the server == retry in background job (retry delay should have some randomness)
  - [ ] Invalid json from server == just fail
  - [ ] Does Sentry catch crashes in PocketWorker?
  - [ ] What happens when exceptions are thrown inside the PocketWorker?

### Impact
Some unused code needs to be removed and this background fetch functionality needs to be tested.
### Acceptance criteria
All the above tasks are completed.

**Comment by npark-mozilla** [comment]
Just my 2 cents:

So this is my understanding of the background fetch works (which could be wrong):

- Background fetch is scheduled typically between anytime during 3AM and 6AM

- When the app is started, it checks for the fetched video JSON, and schedules the next fetch time
- Initially there is a hardcoded video JSON (so there is always something to show on the overlay)
- Assuming it will always fetch the list of video from pocket server that’s different from what we currently have 
- We don’t have the record of the last successful fetch timestamp

If above is correct, it would be ideal if we have following grey-box tests that does the following, in addition to the usual unit tests. They cover pretty typical scenarios:

Checking whether update actually happens nominally:
- Opens app, checks the fetch is done, checks the next update schedule value is set.
- Change the next update schedule to currentTime + 1000ms
- wait to see whether pocketstore updates and refreshes overlay when reopened

Checking whether pocket updates past the specified time:
- Opens app, checks the fetch is done, checks the next update schedule value is set.
- Change the next update schedule to currentTime or past time (assuming this gets noticed by the app)
- Check whether the pocketstore updates (since the scheduled time is now in the past) and refreshes overlay when reopened

Checking whether app does not crash during network is down, and schedule reattempt 
- Opens app, checks the fetch is done, checks the next update schedule value is set.
- Disable wifi
- Change the next update schedule to currentTime + 1000ms
- wait too see the app does not crash after failed fetch attempt, and check backoff time is properly set

Alternatively, above scenario can be done manually too, but then we need a way for the manual tester to view/edit next update schedule via about:config or so. Also, I think different locale test can be covered manually too.  If we go that route, I'll update TestRail with additional test cases.

**Comment by npark-mozilla** [comment]
For the cases of:
- If a user's device is turned off and turned back on before the scheduled time, does WorkManager still run the job at the scheduled time? => Yes
- If the device is turned off at the scheduled time and turned on later, does WorkManager still run the job (e.g. it catches up on old work)? => No
I have tested with the local build, and above are what I found

**Comment by mcomella** [comment]
Thanks for testing, No-Jun!

No-Jun suggested some thorough test cases in https://github.com/mozilla-mobile/firefox-tv/issues/2325#issuecomment-496566537. We wrote unit tests to cover each component but did not cover the integration (i.e. making sure MainActivity calls into our components: will schedule background fetch jobs & push UI updates). We have a few options to address this:
- Write unit tests that cover the integration points: Robolectric throws hard-to-fix errors when we try to create an Activity so this isn't feasible
- Write on-device integration test: these are slow, take a long time to write, and are hard to maintain; tbh, I'd rather have spent the time getting the unit tests to work
- Expose functionality in production for QA to test: we went with this given the issues below; our current strategy is to add flags to Intents

**Comment by mcomella** [comment]
dnarcese is taking over the remaining work so unassigning myself.

**Comment by mcomella** [comment]
> We wrote unit tests to cover each component but did not cover the integration

To be explicit, if we had wanted to cover the integration with unit tests, we should have additionally tested:
- Does MainActivity call into the scheduler to add the job?
- Does MainActivity call into the VideoRepo to push state in onStart?

**Comment by dnarcese** [comment]
>  - Does Sentry catch crashes in PocketWorker? + rm TODO in PocketVideoFetchWorker
>  - What happens when exceptions are thrown inside the PocketWorker? + rm TODO in PocketVideoFetchWorker

I forced a crash in `doWork`  (by `Sentry.capture`) and what I found is that:
- it does not crash the app
- if the app is backgrounded, Sentry captures it
- if the process is killed, `doWork` is never hit so there is no Sentry error

**Comment by mcomella** [comment]
On Mon, Jun 3, 2019 at 3:20 PM Daniela Arcese <notifications@github.com>
wrote:

> I forced a crash in doWork (by Sentry.capture) and what I found is that:
>
@dnarcese Did you have the opportunity to revisit this by throwing an
Exception as we discussed on Slack? I'm just following up to make sure I
was explicit enough.

**Comment by dnarcese** [comment]
@mcomella Yes! I updated the comment with my results of using `throw RuntimeException`

**Comment by mcomella** [comment]
> @mcomella Yes! I updated the comment with my results of using `throw RuntimeException`

Keep in mind we don't get email updates for edits! :)

If that's the case, can you file a follow-up to investigate how we might get Sentry working in WorkManager?

**Comment by dnarcese** [comment]
Completed in [b8b29d](https://github.com/mozilla-mobile/firefox-tv/commit/b8b29d90561232fc8dfb77657f92d85bc0aee5f3).
